_posts/function-contracts.md

```diff
@@ -0,0 +1,298 @@
+Verification is a costly art. If you are an active user of Kani you know that the runtime of a proof does not scale linearly with the amount of code under verification and you probably have a number of functions and ideas for harnesses you’d like Kani to verify, but it just takes prohibitively long.
+
+We [previously](https://model-checking.github.io/kani-verifier-blog/2023/02/28/kani-internship-projects-2022-stubbing.html) blogged about stubbing, a technique where a function that performs a costly computation, for instance trough recursion or a long running loop, could be replaced with a function that is cheaper to execute in the verifier. This replacement approximates the behavior of the function it replaces and/or uses techniques only available in a verifier, like non-deterministic values (`kani::any`). By successively replacing costly functions with stubs Kani can check ever larger proofs but it also comes with a downside.
+
+As we discussed in the [section on risks](https://model-checking.github.io/kani-verifier-blog/2023/02/28/kani-internship-projects-2022-stubbing.html#risks-of-stubbing) stubbing is a powerful tool with few guard rails. In principle you can stub a function with any other function, so long as the type signature matches. Of course any user of stubs is going to be careful to align the behavior of the function being stubbed and the stub. However the whole point of stubbing is that the stub doesn’t behave exactly as the function it replaces, because it needs to be faster. If the approximation is incorrect, it leads to *unsoundness*, meaning that verification can succeed in spite of a bug being present.
+
+In today's blogpost we want to introduce you to a new concept that lets us verify larger programs without sacrificing soundness: function contracts. The idea is very simple: like a stub the contract approximates the behavior of a function in an efficient way, allowing us to scale and verify larger programs. However with a contract the verifier automatically checks that the approximation is sound so we can be confident in the results of our verification.
+
+## Dangers of Stubbing
+
+Let us illustrate this with an example. Consider Euclid’s formula for calculating the greatest common divisor of two numbers, here called `max` and `min`. For verification this would be considered an expensive computation due to the recursive call. This recursion is very busy, in the [worst case](https://en.wikipedia.org/wiki/Euclidean_algorithm#Worst-case) the number of steps (recursions) approaches 1.5 times the number of bits needed to
+represent the number. Meaning that for two large 32 bit numbers it can take almost 48 iterations for a single call to `gcd`. If a harness that uses `gcd` does not heavily constrain the input, Kani would have to unroll the recursion 48 times and then execute it symbolically. An expensive operation.
+
+```rust
+fn gcd(mut max: i32, mut min: i32) -> i32 {
+    if min > max {
+				std::mem::swap(&mut max, &mut min);
```

Comment:
this feels like the names of parameters are not well chosen since you fix the ordering in case it does not match the naming